### ISSUE TITLE
Fastify must not  mutate passed in options

### ISSUE THREAD

**Issue by stalniy** [fastify/fastify#2455]
## 🐛 Bug Report

Fastify mutates passed in options

## To Reproduce

This is mimalistic example to reproduce the issue but the actual issue happens when trying to pass configuration using `config` npm package as options to `fastify` function:

```js
import fastify from 'fastify';

const server = fastify({
  logger: Object.defineProperty({}, 'level', { value: 'info' })
});
```

This code throws:

```
Uncaught TypeError: Cannot assign to read only property 'level' of object '#<Object>'
    at createLogger (/api/node_modules/fastify/lib/logger.js:85:26)
    at fastify (/api/node_modules/fastify/fastify.js:87:33)
```

## Expected behavior

Fastify should not mutate passed in options object

## Your Environment

- *node version*: 12.17.0
- *fastify version*: 3.1.1
- *os*: Linux

**Comment by jsumners** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by stalniy** [comment]
Unfortunately not in the nearest future, I'm pretty occupied by my own OSS projects right now

**Comment by zhangwinning** [comment]
I'm willing to try to work on submit a fix PR.

**Comment by zekth** [comment]
What if instead of: 
https://github.com/fastify/fastify/blob/master/fastify.js#L51-L53
we do this:
```js
function fastify (_options) {
  // Options validations
  let options = Object.assign({}, _options)
```

**Comment by jsumners** [comment]
It should be a deep clone so that Fastify can retain a clean copy. But I think someone had arguments against that at some point; I'm not sure, though.

**Comment by stalniy** [comment]
It can be a shallow clone in places where you to set default values. 


```js
// instead of this 
options.logger.level =options.logger.level || 'info'

// do this
const loggerOptions = { level: 'info', ... options.logger }
```

**Comment by stalniy** [comment]
First of all guys, thanks for your work! But I must to leave my negative feedback about implementation of the fix at line: https://github.com/fastify/fastify/blob/master/lib/logger.js#L86

This doesn’t look better than a deep clone, probably slower, harder to read and not future proof. If somebody will try to set default values for another deeply nested property in the future, we will have the same bug again.

**Comment by zekth** [comment]
> It should be a deep clone so that Fastify can retain a clean copy. But I think someone had arguments against that at some point; I'm not sure, though.

indeed `Object.assign` won't work with versionning for example.

What about rewriting this: 
https://github.com/fastify/fastify/blob/master/fastify.js#L94-L100
and make an inner `option` object and using spread operator when needed like suggested @stalniy ?

**Comment by RafaelGSS** [comment]
> and make an inner option object and using spread operator when needed

I don't see any problem with it. Since most of these properties could be deeply changed in the future, sounds good to make a copy of it (considering that is not a hot path)

**Comment by Eomm** [comment]
This should be fixed in #2474 

@zekth are you suggesting to deepcolne all the `options`?
